plone/app/vocabularies/catalog.py

```diff
@@ -646,6 +665,116 @@ def __call__(self, context, query=None):
+    """Catalog Vocabulary for static queries of content based on a fixed query.
+    Intended for use in a zope.schema, e.g.:
+
+        my_relation = RelationChoice(
+            title="Custom Relation",
+            vocabulary=StaticCatalogVocabulary({
+                "portal_type": "Document",
+                "review_state": "published",
+            })
+        )
+
+    Can be used with TextLine values (to store a UUID) or
+    Relation/RelationChoice values (to create a z3c.relationfield style
+    relation). This vocabulary will work with a variety of selection widgets,
+    and provides a text search method to work with the
+    plone.app.z3cform.widget.AjaxSelectWidget.
+
+    This vocabulary can be used to make a named vocabulary with a factory
+    function:
+
+        from zope.interface import provider
+        from zope.schema.interfaces import IVocabularyFactory
+
+
+        @provider(IVocabularyFactory)
+        def my_vocab_factory(context):
+            return StaticCatalogVocabulary({
+                'portal_type': 'Event',
+                'path': '/'.join(context.getPhysicalPath())
+            })
+
+    The default item title looks like "Object Title (/path/to/object)", but this
+    can be customized by passing a format string as the "title_template"
+    parameter. The format string has "brain" and "path" arguments available:
+
+        MY_VOCABULARY = StaticCatalogVocabulary(
+            {'portal_type': 'Event'},
+            title_template="{brain.Type}: {brain.Title} at {path}"
+        )
+
+    When using this vocabulary for dynamic queries, e.g. with the
+    AjaxSelectWidget, you can customize the index searched using the
+    "text_search_index" parameter. By default it uses the "SearchableText"
+    index, but you could have your vocabulary search on "Title" instead:
+
+        from plone.autoform import directives
+        from plone.app.z3cform.widget import AjaxSelectFieldWidget
+
+
+        directives.widget(
+            'my_relation',
+            AjaxSelectFieldWidget,
+            vocabulary=StaticCatalogVocabulary(
+                {'portal_type': 'Event'},
+                text_search_index="Title",
+                title_template="{brain.Type}: {brain.Title} at {path}"
+            )
+        )
+
+    This vocabulary lazily caches the result set for the base query on the
+    request to optimize performance.
+    """
+    title_template = "{brain.Title} ({path})"
+    text_search_index = "SearchableText"
+
+    def __init__(self, query, text_search_index=None,
+                 title_template=None):
+        self.query = query
+        if text_search_index:
+            self.text_search_index = text_search_index
+        if title_template:
+            self.title_template = title_template
+
+    @staticmethod
+    def get_request():
+        return getRequest()
+
+    @property
+    @request.cache(get_key=request_query_cache_key, get_request="self.get_request()")
+    def brains(self):
```

Comment:
After https://github.com/plone/plone.memoize/pull/22/files you should be able to achieve the same using the plone.memoize.view.memoize decorator